src/inmanta/server/agentmanager.py

```diff
@@ -213,108 +179,92 @@ async def _unpause_agent(self, env: data.Environment, endpoint: Optional[str] =
                 live_session = self.tid_endpoint_to_session.get(key)
                 # If the agent has a live_session, the agent wasn't paused
                 if not live_session:
-                    session = self._get_session_for(tid=env.id, endpoint=agent_name)
+                    session = self.get_session_for(tid=env.id, endpoint=agent_name)
                     if session:
                         self.tid_endpoint_to_session[key] = session
                         await session.get_client().set_state(agent_name, enabled=True)
 
-    # Agent Management
-    async def ensure_agent_registered(self, env: data.Environment, nodename: str) -> data.Agent:
-        """
-            Make sure that an agent has been created in the database
-        """
-        async with self.session_lock:
-            agent = await data.Agent.get(env.id, nodename)
-            if agent is not None:
-                return agent
-            else:
-                return await self._create_default_agent(env, nodename)
+    # Notify from session listener
+    async def new_session(self, session: protocol.Session) -> None:
+        await self._register_session(session, datetime.now())
 
-    async def _create_default_agent(self, env: data.Environment, nodename: str) -> data.Agent:
-        """
-            This method creates a new agent (agent in the model) in the database.
-            If an active agent instance exists for the given agent, it is marked as a the
-            primary instance for that agent in the database.
+    # Notify from session listener
+    def expire(self, session: protocol.Session, timeout: float) -> None:
+        self.add_background_task(self._expire_session(session, datetime.now()))
```

Comment:
```suggestion
        # Expiry, creation and endpoint update (for different session) will come in here in-order, but could get re-ordered by the add_background_task. This is safe because all scenarios result in the same end-result: the primary is moved to the newly created session. 
        self.add_background_task(self._expire_session(session, datetime.now()))
```